File: frontend/mobile_app/lib/features/cart/cart_controller.dart
import 'package:flutter/foundation.dart';

import '../../models/cart.dart';
import '../../models/product.dart';

/// Owns the cart, which in this showroom doubles as the salesperson's
/// **shortlist and on-screen selector**. Adding an item optionally presents it;
/// tapping a cart item re-presents it on the display (wired in the UI layer).
class CartController extends ChangeNotifier {
  Cart _cart = const Cart();
  Cart get cart => _cart;

  void addItem(
    Product product, {
    required String variantId,
    required String size,
  }) {
    final CartItem incoming = CartItem(
      product: product,
      variantId: variantId,
      size: size,
    );
    final List<CartItem> items = List<CartItem>.of(_cart.items);
    final int idx = items.indexWhere(
      (CartItem i) => i.lineId == incoming.lineId,
    );
    if (idx >= 0) {
      items[idx] = items[idx].copyWith(quantity: items[idx].quantity + 1);
    } else {
      items.add(incoming);
    }
    _cart = _cart.copyWith(items: items);
    notifyListeners();
  }

  void setQuantity(String lineId, int quantity) {
    final List<CartItem> items = List<CartItem>.of(_cart.items);
    final int idx = items.indexWhere((CartItem i) => i.lineId == lineId);
    if (idx < 0) return;
    if (quantity <= 0) {
      items.removeAt(idx);
    } else {
      items[idx] = items[idx].copyWith(quantity: quantity);
    }
    _cart = _cart.copyWith(items: items);
    notifyListeners();
  }

  /// Change the size of an existing line. Because a line's identity includes its
  /// size, switching to a size that already exists in the cart merges the two
  /// lines (summing quantities); otherwise the line is re-sized in place.
  void setSize(String lineId, String size) {
    final List<CartItem> items = List<CartItem>.of(_cart.items);
    final int idx = items.indexWhere((CartItem i) => i.lineId == lineId);
    if (idx < 0) return;
    final CartItem resized = items[idx].copyWith(size: size);
    if (resized.lineId == lineId) return; // unchanged
    final int existing = items.indexWhere(
      (CartItem i) => i.lineId == resized.lineId,
    );
    if (existing >= 0) {
      items[existing] = items[existing].copyWith(
        quantity: items[existing].quantity + resized.quantity,
      );
      items.removeAt(idx);
    } else {
      items[idx] = resized;
    }
    _cart = _cart.copyWith(items: items);
    notifyListeners();
  }

  void removeItem(String lineId) {
    final List<CartItem> items = _cart.items
        .where((CartItem i) => i.lineId != lineId)
        .toList();
    _cart = _cart.copyWith(items: items);
    notifyListeners();
  }

  void clear() {
    _cart = const Cart();
    notifyListeners();
  }
}

File: frontend-rn/mobile_app/src/features/cart/cartController.ts
import { Listenable } from '../../core/listenable';
import { Cart, CartItem } from '../../models/cart';
import { Product } from '../../models/product';

/// Owns the cart, which doubles as the salesperson's shortlist and on-screen
/// selector. Ported 1:1 from `CartController`.
export class CartController extends Listenable {
  private _cart = new Cart();
  get cart(): Cart {
    return this._cart;
  }

  addItem(product: Product, opts: { variantId: string; size: string }): void {
    const incoming = new CartItem({ product, variantId: opts.variantId, size: opts.size });
    const items = [...this._cart.items];
    const idx = items.findIndex((i) => i.lineId === incoming.lineId);
    if (idx >= 0) {
      items[idx] = items[idx].copyWith({ quantity: items[idx].quantity + 1 });
    } else {
      items.push(incoming);
    }
    this._cart = this._cart.copyWith({ items });
    this.notifyListeners();
  }

  setQuantity(lineId: string, quantity: number): void {
    const items = [...this._cart.items];
    const idx = items.findIndex((i) => i.lineId === lineId);
    if (idx < 0) return;
    if (quantity <= 0) {
      items.splice(idx, 1);
    } else {
      items[idx] = items[idx].copyWith({ quantity });
    }
    this._cart = this._cart.copyWith({ items });
    this.notifyListeners();
  }

  /// Change the size of an existing line — merges with an existing same-size line.
  setSize(lineId: string, size: string): void {
    const items = [...this._cart.items];
    const idx = items.findIndex((i) => i.lineId === lineId);
    if (idx < 0) return;
    const resized = items[idx].copyWith({ size });
    if (resized.lineId === lineId) return; // unchanged
    const existing = items.findIndex((i) => i.lineId === resized.lineId);
    if (existing >= 0) {
      items[existing] = items[existing].copyWith({ quantity: items[existing].quantity + resized.quantity });
      items.splice(idx, 1);
    } else {
      items[idx] = resized;
    }
    this._cart = this._cart.copyWith({ items });
    this.notifyListeners();
  }

  removeItem(lineId: string): void {
    this._cart = this._cart.copyWith({ items: this._cart.items.filter((i) => i.lineId !== lineId) });
    this.notifyListeners();
  }

  clear(): void {
    this._cart = new Cart();
    this.notifyListeners();
  }
}
